### ISSUE TITLE
net share does not work since Windows Server 2012 R2

### ISSUE THREAD

**Issue by shino0729** [dokan-dev/dokany#200]
# Environment
 * Windows version: Windows Server 2012 R2
 * Processor architecture: x64
 * Dokany version: v1.0.0-RC2
 * Library type (Dokany/FUSE): Dokany
 
# Check List
  - [x] I checked my issue [doesn't exist yet](https://github.com/dokan-dev/dokany/issues?utf8=%E2%9C%93&q=is%3Aissue)
  - [x] My issue is valid with mirror default sample and not specific to my user-mode driver implementation
  - [x] I can always reproduce the issue with the provided description bellow
  - [x] If I updated Dokan driver (installer or .sys), I tried to restart the computer before my tests
  - [x] It can be reproduced on latest [release and pre-release](https://github.com/dokan-dev/dokany/releases)

# Description
Cannot access to the shared mount point made by the mirror sample.

    > mirror.exe /r c:\tmp /l m
    > net share <share_name>=m:\ /grant:Everyone,full

Then access to the shared directory (\\\\127.0.0.1\\\<share_name\>), return the error "device is not ready".
 
On Windows Server 2008 R2, it works fine.

**Comment by js69** [comment]
Just to add my observation, as net sharing a Dokan drive never worked for me on Win 8 and WIn 10: tested it now on Win7-x86 and it works there.

**Comment by Liryna** [comment]
This is the job of the network provider (dokannp1.dll).
You can find the source here: https://github.com/dokan-dev/dokany/blob/master/dokan_np/dokannp.c

It is a very simple implementation and it is probably now outdated for new systems after Win7 as you describe.

I think it need to be reviewed/rewrite.

**Comment by js69** [comment]
Could we learn something from the VirtualBox implementation of a network provider?

http://www.virtualbox.org/svn/vbox/trunk/src/VBox/Additions/WINNT/SharedFolders/np/vboxmrxnp.cpp

**Comment by Liryna** [comment]
@shino0729 do you have the possibility to change registry:
[HKEY_LOCAL_MACHINE\SYSTEM\ControlSet001\Control\NetworkProvider\HwOrder]
[HKEY_LOCAL_MACHINE\SYSTEM\ControlSet001\Control\NetworkProvider\Order]

Rename in ProviderOrder DokanNP1 to Dokan1 and make your test again ?

I am currently not able to use net share on my VM:
```
>net share test=m:\ /grant:MyUser,FULL
The operation is invalid on a redirected resource.
```
Am I doing something wrong ?

**Comment by shino0729** [comment]
Hi @LirynaI , I changed the registry but I couldn't access to the net shared folder.

net share:
It seems like using network drive. Did you execute mirror sample  without **/n** option?

**Comment by Liryna** [comment]
>< Ok I did without /n and it worked on my windows 10 but I face the same issue as you :cry: 
I am not sure this is related to network provider therefore.

Did you try to access directly a subfolder : \127.0.0.1\<share_name>\test ?

@Maxhy If I remember there is a trick here like with the UNC  ?

**Comment by shino0729** [comment]
Yes, I tried to access directly a subfolder/file, but I couldn't access.

**Comment by Maxhy** [comment]
I think it never worked on Win8/8.1/10. I remember this is an old issue even if it wasn't created on Dokany issue tracker until now, there is similar reports on archived google code issue tracker https://code.google.com/archive/p/dokan/issues/302
Need to be fixed.

**Comment by Maxhy** [comment]
Ok I accidentally found something related to this issue during another test.
If Filetrace (from FileSpy) mini filter driver is attached to the volume, then sharing will work... It's a side effect but it looks like something is not correctly initialized until it got attached.
Any hints would be helpful. @marinkobabic? @js69? @Liryna?

**Comment by marinkobabic** [comment]
@Mathy
How did you your tests: 

Did you use the network drive option and the unc option during the mount? 
Did you use a fixed drive and then you shared it and tried it to acces using unc path?
In generally you tried to access the drive share from same machine or from another machine? 
You used unc path or did you mount the share before? 

Thanks
Marinko

**Comment by Liryna** [comment]
I made a test @Maxhy with Filespy and have not been able to make it work even after attaching it :cry: 

From my test:
```
> mirror.exe /r c:\Users /l m
> net share test=m:\ /grant:Everyone,full
```
The device is 'well' shared. You can see it even from a remote computer without issue.
BUT when you try to acces it from local (\\127.0.0.1\test ) or remote (\\NAME\test ), it does not work.
You have an error "device is not ready".
Mirror logs: http://pastebin.com/YPnqdbhz

With UNC:
```
> mirror.exe /r c:\Users /l m /n /u \localhost\mytest
```
It work perfectly from local with `\\localhost\mytest`.
When UNC is used, it is only available from local computer even if UNC used is `/u \NAME\mytest`

**Comment by Liryna** [comment]
@marinkobabic I have been able to reproduce what @Maxhy explained.
Mirror need to use the mount manager option `/o` for making it work with filespy.

```
> mirror.exe /r c:\Users /l m /o
> net share test=m:\ /grant:Everyone,full

//attach with FileSpy
-> \\localhost\mytest - WORK !
```

**Comment by Liryna** [comment]
Logs:

Access without FileSpy attached (error): http://pastebin.com/pzSyEkix
FileSpy Attaching: http://pastebin.com/bwDJ62JD
Access with FileSpy attached (success): http://pastebin.com/jSycWhpU

**Comment by Deathspike** [comment]
Is there any additional information on this issue? I don't mind spending some time digging through Dokan for a fix for this, but I don't know where to look. Any kind of additional information/guesses would be highly appreciated.

**Comment by Liryna** [comment]
Hi @Deathspike ,

I do not have much more information unfortunately. My guess would be that one of the request made by the system to the kernel is wrong/denied/not supported and the system does not like it.

Since it only happen on 2012 R2, I would guess that it is an unsupported new request.

If you need any help on it, feel free to ask !

**Comment by siegelpeter** [comment]
Hi!

Perhaps i can help. I got the same problem on Windows 8.1 N 64 bit. I don't
think its limited to windows server 2012 R2.

Are there any logs which would be helpfull to you?



2016-06-27 13:34 GMT+02:00 Liryna <notifications@github.com>:

**Comment by Liryna** [comment]
Hi @siegelpeter ,

Windows server 2012 R2 and Windows 8 have the same kernel, thats why you also had the issue.

The only possibility is to dig in what the system request or to know what microsoft changed in windows 8 behaviour for net share.

**Comment by Deathspike** [comment]
Yeah, it happens since Windows Server 2012 R2. That is NT6.3+. That includes Windows 8.1 and Windows 10. I think it would be useful to see the differences in logging between Windows 7/8 and 8.1. @Liryna posted logs before and after attaching FileSpy, which may prove useful too, but if FileSpy is hooking into things, the actual difference may be obscured. A change log would be good, too, but I doubt we'll find that.

**Comment by siegelpeter** [comment]
i have windows 7 on a vm in the office,
I can try to get logs for you for windows 7 too,

In my case attaching filespy made no diference as far as i could tell.



2016-06-27 14:55 GMT+02:00 Roel van Uden <notifications@github.com>:

**Comment by Deathspike** [comment]
It would help!

If we can compare a dump from Windows 7 and Windows 8.1+, I suspect we can put them side by side and compare for differences. It probably won't give an immediate answer, but it should give a good indication which (first) step fails. At least I will know where to start digging. 

So if you could get a log of both systems doing exactly the same:
* Windows 7: mount, access via network, show empty folder (because less logging), stop logging.
* Windows 8: mount, access via network, error, stop logging.

That would be awesome :+1:

**Comment by Maxhy** [comment]
Fixed with b6060d437b975b712432412c1cbc6f8e0298d4bc thanks to @Kerbox!
